sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/DynamicDestinations.java

```diff
@@ -0,0 +1,166 @@
+ *
+ * <p>An instance of {@link DynamicDestinations} can also request a side input value that can be
+ *  examined from inside {@link #getTable} and {@link #getSchema}. The side input is requested by
+ *  calling {@link #setSideInputRequired} on the base class. The value can be examined via the
+ *  {@link DynamicDestinations.SideInputAccessor} class.
+ *
+ * <p>{@code DestinationT} is expected to provide proper hash and equality members. Ideally it will
+ * be a compact type with an efficient coder, as these objects may be used as a key in a
+ * {@link org.apache.beam.sdk.transforms.GroupByKey}.
+ */
+public abstract class DynamicDestinations<T, DestinationT> implements Serializable {
+  private PCollectionView<?> sideInput;
+
+  /**
+   * Returns the materialized value of the side input. Can be used by concrete
+   * {@link DynamicDestinations} instances in {@link #getSchema} or {@link #getTable}.
+   */
+  public static class SideInputAccessor {
+    private ProcessContext processContext;
+    private PCollectionView<?> sideInput;
+    SideInputAccessor(ProcessContext processContext, PCollectionView<?> sideInput) {
+      this.processContext = processContext;
+      this.sideInput = sideInput;
+    }
+
+    public <SideInputT> SideInputT getSideInputValue() throws IllegalStateException {
+      if (sideInput == null) {
+        return null;
+      }
+      @SuppressWarnings("unchecked")
+      SideInputT materialized = (SideInputT)  processContext.sideInput(sideInput);
+      return materialized;
+    }
+  }
+
+  /**
+   * Specifies that this object needs access to a side input. This side input must be globally
+   * windowed, as it will be accessed from the global window.
+   */
+  public DynamicDestinations setSideInputRequired(PCollectionView<?> sideInput) {
+    this.sideInput = sideInput;
+    return this;
+  }
+
+  /**
+   * Returns an object that represents at a high level which table is being written to.
+   */
+  public abstract DestinationT getDestination(ValueInSingleWindow<T> element);
+
+  /**
+   * Returns the coder for {@link DestinationT}. If this is not overridden, then
+   * {@link BigQueryIO} will look in the coder registry for a suitable coder. This must be a
+   * deterministic coder, as {@link DestinationT} will be used as a key type in a
+   * {@link org.apache.beam.sdk.transforms.GroupByKey}.
+   */
+  @Nullable
+  public Coder<DestinationT> getDestinationCoder() {
+    return null;
+  }
+
+  /**
+   * Returns a {@link TableDestination} object for the destination.
+   */
+  public abstract TableDestination getTable(DestinationT destination,
+                                            SideInputAccessor sideInputAccessor);
+
+  /**
+   * Returns the table schema for the destination.
+   */
+  public abstract TableSchema getSchema(DestinationT destination,
+                                        SideInputAccessor sideInputAccessor);
+
+  /**
+   * This returns the unmaterialized side input used by this transform.
```

Comment:
I'd think that with SideInputAccessor, we no longer need to assume that there's only one side input.
SideInputAccessor can be an interface with a method `public <T> T sideInput(PCollectionView<T> view)`, and DynamicDestinations can have a `List<PCollectionView<?>> getSideInputs()`. Then you can get rid of the ugly `setSideInputRequried`